PythonVisualizations/PriorityQueue.py

```diff
@@ -0,0 +1,438 @@
+        callEnviron |= set(indexLo)
+        hi = len(self.list)-1              # Point to hi
+        indexHi = self.createIndex(hi, 'hi', level = 3)
+        callEnviron |= set(indexHi)
+        mid = (lo + hi) // 2               # Point to the midpoint
+        indexMid = self.createIndex(mid, 'mid', level=2)
+        callEnviron |= set(indexMid)
+        while lo <= hi:
+            mid = (lo + hi) // 2           # Select the midpoint
+            if self.list[mid].val == val:  # Did we find it at midpoint?
+                self.stopAnimations()
+                self.window.update()
+                return mid                 # Return the value found
+
+            elif self.list[mid].val < val: # Is item in upper half?
+                deltaXLo = (mid - lo) + 1
+                self.moveItemsBy(indexLo, (self.CELL_SIZE*deltaXLo, 0))
+                lo = mid + 1               # Yes, raise the lo boundary
+                deltaXMid = ((hi - lo) // 2) + 1
+                self.moveItemsBy(indexMid, (self.CELL_SIZE*deltaXMid, 0))
+
+            else:                         # Is item in lower half?
+                deltaXHi = (mid-hi) - 1
+                self.moveItemsBy(indexHi, (self.CELL_SIZE*deltaXHi, 0))
+                hi = mid - 1              # Yes, lower the hi boundary
+                deltaXMid = ((lo-hi)//2)-1
+                self.moveItemsBy(indexMid, (self.CELL_SIZE*deltaXMid, 0))
+
+        self.stopAnimations()
+        self.window.update()
+        self.cleanUp(callEnviron)
+        return lo                         # val not found
+
+    # delete the last element of the queue, or None if empty
+    def remove(self):
+        callEnviron = self.createCallEnvironment()
+        self.startAnimations()
+        found = len(self.list) > 0
+        if found:  # Record if value was found
+            self.wait(0.3)
+
+            n = self.list[len(self.list) - 1]
+
+            indexEnd = self.createIndex(len(self.list) - 1, 'front', level=-1)  # indicate priority
+            callEnviron |= set(indexEnd)
+
+            # Slide value rectangle up and off screen
+            items = (n.display_shape, n.display_val)
+            self.moveItemsOffCanvas(items, N, sleepTime=0.02)
+
+            self.moveItemsBy(indexEnd, (-self.CELL_SIZE, 0))  # move priority arrow
+
+            self.removeFront()
+
+        self.cleanUp(callEnviron)
+        return found
+
+    def removeFront(self):
+        callEnviron = self.createCallEnvironment()
+
+        # pop a Drawable from the list
+        if len(self.list) == 0:
+            self.setMessage('Array is empty!')
+            return
+        self.startAnimations()
+        n = self.list.pop()
+
+        # delete the associated display objects
+        self.canvas.delete(n.display_shape)
+        self.canvas.delete(n.display_val)
+
+        # update window
+        self.window.update()
+        self.cleanUp(callEnviron)
+        self.stopAnimations()
```

Comment:
The PriorityQueue only needs one Remove operation.  That always removes the item at the front of the queue (the one with the highest value and at the highest array index in this implementation).  The `nItems` pointer is updated as a result.   This second `removeFront()` method should be removed.